[1.15.2] Allow ForgeConfigSpec to better support strings and lists

Senmori opened the pull request:
This PR allows ForgeConfigSpec to better support strings and lists in configuration files.

This introduces two new `ConfigValue` subclasses. `StringValue` and `ListValue<T>`.

The problem with the current implementation is that Strings are not supported as well as the primitive values.
This would allow users to migrate from `ConfigValue<String>` to `StringValue`.

Lists are the primary reason for this PR. List support in `ConfigValue` is a bit lacking.
This allows list elements to be converted to their proper type via a given converter.

Unfortunately, I had to put `defineString` instead of just using `define` because of type erasure in the suppliers/lists. I still think this is acceptable because it's still clear as to what the user is defining.

Proof of test working: https://i.imgur.com/abuwqeC.png

Funwayguy committed:
Re-added patch to enable stencil buffer in net.minecraft.client.shader.Framebuffer
Add test for the stencil buffer feature. Only tests that it doesn't crash when enabling, for now. - giga

Senmori committed:
Initial work for dimension deletion PR

Senmori committed:
Add support for strings and lists to ForgeConfigSpec.

Senmori committed:
Because I stashed all changes like an idiot these showed up.

Senmori committed:
Missed this entry in the last commit.

Senmori committed:
Better test to reflect config loading values from toml.

Senmori committed:
Add javadocs to other define methods in ForgeConfigSpec.

BernardoCBranco committed:
Remove ArrowEntity patch, vanilla fixed MC-107941

tterrag1098 committed:
Deprecate ModelBuilder#gui3d, add ModelBuilder#guiLight

tterrag1098 committed:
Add support for optional tag values to Tag.Builder

Add optional tag values to data gen test
Fix data gen test not running by correcting gui_light value

cpw committed:
Add a progress message for atlas textures. Fills a big time gap on bigger packs.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

cpw committed:
Fix cycle happening in noteblock even though we're setting the new value as well..

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

hnOsmium0001 committed:
Fix ownership leak in ItemStackHandler (#6580)

- Triggered by invoking ItemStackHandler#extractItem with `simulate=true`

dags- committed:
Defer writing the server.properties file until after mods have loaded to prevent config loss (#6545)

naqaden committed:
Add renderer to RenderNameplateEvent (#6562)

tterrag1098 committed:
Fix a few issues with BufferBuilder

Fix vanilla bug that caused the byte order of the buffer from getNextBuffer to be incorrect
Fix missing modification to nextElementBytes in putBulkData

tterrag1098 committed:
Move global loot test data files out of generated folder

LexManos committed:
Cleanup IForgeTagBuilder code style.

tterrag1098 committed:
Remove SideOnly from WoodType methods

tterrag1098 committed:
Revive forge lighting pipeline, disabled by default for now

Remove a lot of light value convolution by using 0..1 for nearly everything
Fix a lot of TODOs that are needed for the lighting pipeline to function
Potential fix for #6425

tterrag1098 committed:
Improve performance of VertexBufferConsumer by inlining element locs

tterrag1098 committed:
Fix Tag serializing empty optionals, improve RegistryObject error msg

valoeghese committed:
Added hook to allow modded Hill Biomes (#6571)

patrick-aoc committed:
Fixed PlantType not extending IExtensibleEnum, Closes #6286

tterrag1098 committed:
Fix element type of scanned child annotations

tterrag1098 committed:
Fix child annotations not inheriting member name, cleanup

tterrag1098 committed:
Fix VariantList calling deprecated bake method

gigaherz committed:
Add "type" field to global loot modifier jsons. (#6554)

* Add "type" field to global loot modifier jsons and fix tests.
To avoid binary breaking, "type" defaults to be the same as the filename.
* Fix license headers.

LexManos committed:
Fixed dev time compilation using J9+ producing incorrect bytecode.

LexManos committed:
Revert defualt jvm args.

cpw committed:
Fix recursion mess in PlayerAdvancement loading, by using an alternative based on a toposorted list of all advancements.
It seems to be fully equivalent in testing, but there may be corner cases, so there is a config to disable, though disabling
may break servers, especially those chunkloading fake players.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

gigaherz committed:
Add tickRate setter to the ForgeFlowingFluid Properties builder.

LexManos committed:
Fix potential NPEs in RegistryObject.orElseThrow/isPresent/ifPresent Closes #6647

cpw committed:
Update jenkinsfile and gradle for new versions of things..

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

cpw committed:
merge early startup into main window if used. It seems to work well.. Also ticks window on main thread, so mac compatible now.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

RogueLogix committed:
move handOffWindow down near start()

RogueLogix committed:
Move ClientVisualization to secondary thread

LexManos committed:
Fix Automatic Event Subscriber not detecting mod id defaults, and fix test mods.

LexManos committed:
Fix resource packs not being sorted properly. Closes #6287

cpw committed:
Handle message display a bit better, try to make the UI tick a bit better.. Still a problem with
a modelbake right at the end, happening on the window flip.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

cpw committed:
Use forge logo as short term window Icon, fix Buffer cast problems for new JDK stupid.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

tterrag1098 committed:
Rename window icon image to avoid conflict, fix not using vanilla icon

cpw committed:
Remove silly icon override. Squirrels are fun y'all..

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

diesieben07 committed:
Fix BackgroundScanHandler crashing on servers (#6660)

tterrag1098 committed:
Fix various issues with the loading screen and datagen

Fix NPE on datagen startup
Fix early loading GUI hanging and never terminating after datagen
Add loading GUI messages for datagen

gigaherz committed:
Remove redundant call to modifyLoot. Closes #6657

ichttt committed:
Fix config comments referencing disallowed enum values. (#6542)

williewillus committed:
Allow colors of tooltip to be specified in GuiUtils.drawHoveringText tooltip methods (#6579)

Desharin committed:
Added missing localization for swim speed, and reach distance attributes (#6585)

CAS-ual-TY committed:
Fix milk buckets removing potion effects without filters. (#6609)

Darkhax committed:
Change recipe condition log level to debug. (#6621)

Darkhax committed:
Fix vanilla loot table resource leak. (#6622)

LordMonoxide committed:
Fixed ChunkDataEvent.Save being fired with a null world if chunk isn't fully generated when unloaded. (#6628)

pupnewfster committed:
Fixed Hoppers not inserting fully into custom contains with slots that can hold more then max stack (#6631)

Shadows-of-Fire committed:
Allow overriding of tool level requirements when harvesting a block, via the HarvestCheck event.  (#6632)

smith61 committed:
Fix issue with providing a null ItemStack to onPlayerDestroyItem (#6633)

gigaherz committed:
Avoid discarding the bottom 4 bits of the lightmap coords. It turns out vanilla relies on those bits. Fixes some blocks being darker than they should be in certain circumstances.

gigaherz committed:
Patch the bucket logic to not hardcode some checks to WATER. Fixes not being able to place custom fluids into modded fluid-loggable blocks that accept fluids other than water.

jamierocks committed:
Fix ModListScreen escape key not matching done button behavior. Closes #6672

jamierocks committed:
Fixed informational /forge commands being repeated to all ops when logging is enabled. Closes #6330

cpw committed:
Update to newer eventbus, with typetools compatibility with jvms up to j14. No guarantees on actual game compatibility however.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

covers1624 committed:
Fix IForgeBlock.removedByPlayer not firing on the client. (#6638)

LexManos committed:
Fix infinite loading screen when resources error. Closes #6592 & #6593

ShetiPhian committed:
Fixed FluidTank.drain calling onContentsChanged when simulating (#6684)

jamierocks committed:
Fix vanilla screens escape key not matching done button behavior (#6682)

jamierocks committed:
Fixed CropGrowEvents not firing for Bamboo (#6680)

TBoshoven committed:
Fixed tile entities persisting when switching between states that don't have TEs without changing the block. (#6674)

JaisDK committed:
Allow Bees to use custom hives. (#6656)

LXGaming committed:
Fix exclusion of coremods not filtering correctly when the mods directory is a symbolic link. (#6600)

ichttt committed:
Fix server config directory remaining locked when integrated server is shut down. (#6644)

LexManos commented on src/main/java/net/minecraftforge/common/ForgeConfigSpec.java:
```diff
@@ -329,41 +364,93 @@ public Object correct(Object value) {
             }, defaultSupplier);
         }
+        /** @see #define(List, Supplier, Predicate, Function) **/
+        public <T> ListValue<T> define(String path, List<T> defaultValue, Function<Object, T> elementConverter) {
```
These should all be defineList, like all the list related siblings

Cyborgmas committed:
Fix LevelChange event not firing after enchanting (#6693)

covers1624 committed:
Remove duplicate post of ChunkEvent.Load. (#6697)

LexManos committed:
Fixed Partivle.onGround not updating correctly due to decompile error. Closes #6595

LexManos committed:
Fixed typo in registry alias writing causing infinite loop. Closes #6689

gigaherz committed:
Model system fixes:
- Add "origin" key to determine what origin to use. This will default to the existing (broken) behaviour to avoid breaking changes.
- Fix interaction of transforms with variant transforms (transformation ordering).
- Fix OBJ loader ignoring "diffuseLighting" attribute from the json.
  This attribute will continue not being used if "ambientToFullbright" is on (default) to avoid breaking changes.

EdwinMindcraft committed:
Use IProperty#getName when generating blockstate jsons (#6582)

Close #6520

TheSilkMiner committed:
Allow Music Discs to be created with SoundEvent delegates. (#6523)

LexManos committed:
Fix incorrect argument passed in RenderPipeline potentially causing crash. #6677

LexManos committed:
Fix FireBlock using flammability instead of spread speed when looking for places to spread. Closes #6646

ichttt committed:
fix test mod and add warning comment to framebuffer (#6698)

LexManos committed:
Update MCP Mappings.

LexManos committed:
Bump version for new RB.

tterrag1098 committed:
Add forge registries for applicable new registry types

Expose constructors for all forge registry entries
Also, remove deprecation on ItemCameraTransforms/TransformType

tterrag1098 committed:
Improve look of notification GUI, improve StartupQuery API

tterrag1098 committed:
Fix server startup query info not including new text

tterrag1098 committed:
Fix typo in new registry fields' names

Note: this is technically a breaking change. If you for some reason added code referencing these two fields in the last few hours, sorry.

cpw committed:
Improve the DistExecutor API and introduce some "almost completely" safe mechanisms to handle sided lambdas.

In general, use sided lambdas safely by embedding them in a separately loaded class. There's a whole twitch stream
dedicated to investigating this problem. See link in DistExecutor if you're interested.

Signed-off-by: cpw <cpw+github@weeksfamily.ca>

LexManos committed:
Add --flat argument to data generators to not create data in mod specific directores.
Useful For Forge's test mods mainly.
Standardized data gen cache file's folder separators.
Added ItemModelProvider to BlockStateProvider for cleaner datagen code.
Added .gitattributes to fix datagen json's line endings correct on windows.

LexManos committed:
Remove dead test classes.

gigaherz committed:
Implement loader for ItemLayerModel.
Add support for specifying which layers are fullbright.
Fix improper generation of model edge. - Thanks to iChun and KnightMiner for the fix suggestions.

Senmori committed:
Update methods and javadocs with the correct names.
Rebased branch onto master.
Also added an additional test for server specific config loading.
